feat(eslint-plugin): [switch-exhaustiveness-check] add allowDefaultCaseMatchUnionMember option (#9954)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.mdx

Reviewer: The line 
```md
Additionally, if a new value is added to the union type, a `default` would prevent the `switch-exhaustiveness-check` rule from reporting on the new case not being handled in the `switch` statement.
```
in the `allowDefaultCaseForExhaustiveSwitch` section should be updated, since this option will address that issue

When a `switch` statement over a union type is exhaustive, a final `default` case would be a form of dead code.

> Additionally, if a new value is added to the union type, a `default` would prevent the `switch-exhaustiveness-check` rule from reporting on the new case not being handled in the `switch` statement.
> If you want `default` to prevent `switch-exhaustiveness-check` when new values ​​are added to the union type, try using the `requireDefaultCaseForUnions` option

Reviewer: how about
```suggestion
Additionally, if a new value is added to the union type and you're using `[considerDefaultExhaustiveForUnions](#considerDefaultExhaustiveForUnions)`, a `default` would prevent the `switch-exhaustiveness-check` rule from reporting on the new case not being handled in the `switch` statement.
```

or just omitting this check now that there is a recourse 🤷

Author: It seemed better to change it like the first suggestion, so we reflected it.

If set to false, this rule will also report when a `switch` statement has a case for everything in a union and _also_ contains a `default` case. Thus, by setting this option to false, the rule becomes stricter.

When a `switch` statement over a union type is exhaustive, a final `default` case would be a form of dead code.

> Additionally, if a new value is added to the union type and you're using `[considerDefaultExhaustiveForUnions](#considerDefaultExhaustiveForUnions)`, a `default` would prevent the `switch-exhaustiveness-check` rule from reporting on the new case not being handled in the `switch` statement.

Reviewer 2: ```suggestion
Additionally, if a new value is added to the union type and you're using [`considerDefaultExhaustiveForUnions`](#considerDefaultExhaustiveForUnions), a `default` would prevent the `switch-exhaustiveness-check` rule from reporting on the new case not being handled in the `switch` statement.
```

Since `value` is a non-union type it requires the switch case to have a default clause only with `requireDefaultForNonUnion` enabled.

### `considerDefaultExhaustiveForUnions`

Defaults to false. Whether a `default` clause will be considered as making a switch statement over a union type exhaustive.

> By default, if a `switch` statement over a union type includes a `default` case, then it's considered exhaustive.

Reviewer: this is the opposite way (since the option defaults to false). By default, every branch will be required to be explicitly handled, and one may opt out by enabling `considerDefaultExhaustiveForUnions`

Author: thank you I think I missed changing it at the same time when changing the default value.

> In a switch case, if you enter the default statement, it can be considered that Exhaustive is satisfied, but if you think that entering the default statement did not solve the exhaustive problem of the union type, please set to option `true`.

Reviewer 2: ```suggestion
By default, if a `switch` statement over a union type includes a `default` case, then it's considered exhaustive.
`allowDefaultCaseMatchUnionMember` enforces explicitly handling every constituent of the union type with their own explicit `case`.
This can be useful if you want to make sure every value added to the union receives explicit handling, with the `default` case reserved for reporting an error.
```

> If set to true, `switch` statement over a union type includes a `default` case is considered exhaustive.
> `considerDefaultExhaustiveForUnions` enforces explicitly handling every constituent of the union type with their own explicit `case`.
> This can be useful if you want to make sure every value added to the union receives explicit handling, with the `default` case reserved for reporting an error.

Reviewer: https://github.com/typescript-eslint/typescript-eslint/pull/9954#discussion_r1801500418 applies to the other lines here too...

```suggestion
If set to true, a `switch` statement over a union type that includes a `default` case is considered exhaustive.
Otherwise, the rule enforces explicitly handling every constituent of the union type with their own explicit `case`.
Keeping this option disabled can be useful if you want to make sure every value added to the union receives explicit handling, with the `default` case reserved for reporting an error.
```

> Examples of additional **incorrect** code for this rule with `{ considerDefaultExhaustiveForUnions: true }`:

Reviewer: ```suggestion
Examples of additional **correct** code for this rule with `{ considerDefaultExhaustiveForUnions: true }`:
```

Author: I thought it would be good to provide detailed example cases, so I added a few cases each correct and incorrect.

Reviewer: Makes sense 👍  I just went ahead and refactored this to our typical tabs setup
